source/DmapSupport.cpp

```diff
@@ -184,37 +184,51 @@ HRESULT GetControllerBaseAddress(PWCHAR deviceName, HANDLE & handle, PVOID & bas
                         g_openDeviceMask |= 1 << i;
 
                         IOControlCode^ IOCTL = ref new IOControlCode(0x423, 0x100, IOControlAccessMode::Any, IOControlBufferingMethod::Buffered);
-                        Buffer^ addressBuffer = ref new Buffer(8);
+                        Buffer^ addressBuffer = ref new Buffer(sizeof(DMAP_MAPMEMORY_OUTPUT_BUFFER));
 
                         create_task(device->SendIOControlAsync(IOCTL, nullptr, addressBuffer))
                             .then([addressBuffer, &controllerAddress, &findCompleted, &hr](UINT32 result)
                         {
-                            // We expect a 4-byte address and a 4-byte length to have been transferred 
-                            // into the address buffer by the I/O operation.
+                            // We expect an 8-byte address and a 4-byte length to have been transferred 
+                            // into the address buffer by the I/O operation.  If this is an old driver, 
+                            // we could get a 4-byte address and a 4-byte length instead.
 
-                            // hr should already be S_OK 
-                            if (result == 8)
+                            if (result < 8)
                             {
-                                hr = S_OK;
+                                hr = E_FAIL;
                             }
-
-                            if (SUCCEEDED(hr))
+                            else
                             {
                                 auto reader = DataReader::FromBuffer(addressBuffer);
-                                UINT32 address = { 0 };
+                                uint64_t address = 0;
                                 int i;
-                                for (i = 0; i < 4; i++)
+
+                                if (result < sizeof(DMAP_MAPMEMORY_OUTPUT_BUFFER))
                                 {
-                                    address = address | (((UINT32)reader->ReadByte()) << (8 * i));
+                                    // If we are talking to an old driver, parse a 4-byte address.
+                                    for (i = 0; i < 4; i++)
```

Comment:
Could the two loops, `i < 4` and `i < 8` become a single loop of `i < sizeof(int)`?